src/cpp/session/modules/SessionUserCommands.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+ * SessionUserCommands.cpp
+ *
+ * Copyright (C) 2009-12 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include "SessionUserCommands.hpp"
+
+#include <boost/bind.hpp>
+
+#include <core/Error.hpp>
+#include <core/Exec.hpp>
+
+#include <r/RExec.hpp>
+#include <r/RSexp.hpp>
+#include <r/RRoutines.hpp>
+
+#include <session/SessionModuleContext.hpp>
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace user_commands {
+
+using namespace rstudio::core;
+
+namespace {
+
+json::Array jsonFromUserCommandResult(SEXP resultSEXP)
+{
+   json::Array jsonData;
+   if (TYPEOF(resultSEXP) != VECSXP)
+      return jsonData;
+   
+   Error error;
+   std::size_t n = r::sexp::length(resultSEXP);
+   
+   for (std::size_t i = 0; i < n; ++i)
+   {
+      json::Object jsonResult;
+      SEXP elementSEXP = VECTOR_ELT(resultSEXP, i);
+      
+      std::string action;
+      error = r::sexp::getNamedListElement(elementSEXP, "action", &action);
+      if (error)
+      {
+         LOG_ERROR(error);
+         return json::Array();
+      }
+      jsonResult["action"] = action;
+      
+      std::vector<int> range;
+      error = r::sexp::getNamedListElement(elementSEXP, "range", &range);
+      if (error)
+      {
+         LOG_ERROR(error);
+         return json::Array();
+      }
+      
+      jsonResult["range"] = json::toJsonArray(range);
+      
+      std::string text;
+      error = r::sexp::getNamedListElement(elementSEXP, "text", &text);
+      if (error)
+      {
+         LOG_ERROR(error);
+         return json::Array();
+      }
+      jsonResult["text"]  = text;
+      
+      jsonData.push_back(jsonResult);
+   }
+   return jsonData;
+}
+
+Error noSuchSymbol(const std::string symbol, const ErrorLocation& location)
```

Comment:
From what I can see it looks like only error codes live in `errc`; error functions generally live in the parent namespace of `errc` and use error categories to report errors consistently.